Refactor addlinks to own class take 3 (follows #21658)

amenk opened the pull request:
### Description

This is a follow up of #21658 - please refer to that one for more information.

It contains two steps

* Refactoring: Extract LinkProcessing functionality
* Change: Allow setting link types via DI (the original reason for this PR) 

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

amenk committed:
Refactoring: Extract Link Processing Logic to own Class "LinkProcessor"

amenk committed:
Import: Allow injecting own link types to LinkProcessor

amenk commented:
@sivaschenko can you review this? 

Again, I don't understand the failed tests, as the reports do not show valuable info or are not available.

sivaschenko commented on app/code/Magento/CatalogImportExport/Model/Import/Product/LinkProcessor.php:
```diff
@@ -0,0 +1,295 @@
+     * TODO: inject via DI
+     * @var array
+     */
+    protected $_linkNameToId = [
```
looks like this property can be removed

amenk replied:
not the property, but the default

amenk replied:
Ah, was doubly there .. removed it

sivaschenko commented on app/code/Magento/CatalogImportExport/Model/Import/Product/LinkProcessor.php:
```diff
@@ -0,0 +1,295 @@
+        Link $resource,
+        int $nextLinkId,
+        array $positionAttrId
+    ): void {
```
This method should be refactored and simplified to reduce cyclomatic complexity, otherwise, it might not pass the static tests

amenk replied:
It is moved from the parent class; there I think it passed the static test I suppose. I'd rather not touch it this time --- can still be refactored later.

sivaschenko commented on app/code/Magento/CatalogImportExport/Model/Import/Product.php:
```diff
@@ -836,7 +845,8 @@ public function __construct(
         DateTimeFactory $dateTimeFactory = null,
-        ProductRepositoryInterface $productRepository = null
+        ProductRepositoryInterface $productRepository = null,
+        LinkProcessor $linkProcessor = null
```
@akaplya can this optional constructor dependency be included in a patch release?

amenk replied:
I think it's a not a B/C breaker

sivaschenko reviewed (changes requested):
Thanks for updates @amenk , please see my review comments

Static tests build does not have a report indeed - relaunched it.

amenk committed:
Remove default values from property, as they are injected via DI

amenk commented:
Wondering if https://testing-service.magento-community.engineering/reports/magento/magento2/pull/23191/79eb9096-95df-45ee-99cf-ab111e939625/26732/SampleDataFunctional/allure-report-ce/index.html#behaviors/3b800d5fe38a970fe497bf12e5c9375c/b4979f4599f0efa2/ is failing by mistake or caused by my PR ...

amenk committed:
Fix initialisation of linkNameToId, unused variable, run phpcbf

amenk committed:
Fix testProductWithLinks failing

amenk committed:
Fix CodeStyle problems

amenk committed:
Fix one code sniffer problem; Add phpcs:disable...

...for code style problem existing before and add according FIXME

amenk committed:
Do not call deprecated code

amenk commented:
All checks are green..

sivaschenko commented:
Great, thanks @amenk . Although, it is not a serious B/C problem, additional optional constructor parameters can have a negative B/C effect in some situations. So we are required to get it approved with a responsible architect. @akaplya can you please take a look at this?

amenk commented:
I just want to get this through before the next merge conflict appears :)

amenk commented:
@akaplya @sivaschenko I also want to understand which kind of negative effects that can be.

sivaschenko commented:
@amenk the example of negative effect:
1. Optional constructor parameter added in 2.3.3
2. Extension created based on 2.3.3 that is relying on the optional parameter added and i.e. change the parameter with DI arguments node
3. That extension is installed on Magento 2.3.1 - as there is not optional parameter - we've got a problem

amenk commented:
Thanks... But we have no other choice to add a new class, right?

amenk commented:
*bump* @sivaschenko @akaplya

Can this pleeeeeease be considered for merging?

amenk commented:
bump @sivaschenko @akaplya

this has tag needs-update but I  don't know which update you are expecting from me?

Can this be progressed ?

amenk commented:
@sivaschenko By the way: The idea to extract it to a new class originally came from @orlangur  in https://github.com/magento/magento2/pull/21230#discussion_r257197555

So I think it's the best way to proceed here. Or is there any other solution for the problem?

sidolov commented:
@amenk you are right,  idea with extraction is the best way here. Could you please cover changes with automated tests?

amenk commented:
@sidolov Thanks for your response. Actually I did mostly make no changes to functionality, just the extraction behind the scenes.
The only functional change is, that the list of link types can be injected via DI. Not sure if that needs to be covered with tests, as DI is a core function.

sidolov commented:
@amenk since you extracted a large piece of code to the new class it would be great to cover it with integration test to ensure that all works as expected. Thanks!

amenk commented:
and again a merge conflict ... but I will try to tackle the problem with the missing tests soon

amenk commented:
The was logic added to delete links which needs to be ported

![conflict1](https://user-images.githubusercontent.com/1087128/75347082-51e20500-58a0-11ea-8508-0f9a3b946de1.png)
![conflict2](https://user-images.githubusercontent.com/1087128/75347089-53abc880-58a0-11ea-9ce7-a5a0e0e364a8.png)

`    /**
     * Delete links
     *
     * @param Link $resource
     * @param array $linksToDelete
     * @return void
     * @throws LocalizedException
     */
    private function deleteProductsLinks(Link $resource, array $linksToDelete)
    {
        if (!empty($linksToDelete) && Import::BEHAVIOR_APPEND === $this->getBehavior()) {
            foreach ($linksToDelete as $linkTypeId => $productIds) {
                if (!empty($productIds)) {
                    $whereLinkId = $this->_connection->quoteInto('link_type_id', $linkTypeId);
                    $whereProductId =  $this->_connection->quoteInto('product_id IN (?)', array_unique($productIds));
                    $this->_connection->delete(
                        $resource->getMainTable(),
                        $whereLinkId . ' AND ' . $whereProductId
                    );
                }
            }
        }
    }`

amenk committed:
Merge remote-tracking branch 'upstream/2.4-develop' into refactor-addlinks-to-own-class-take-3

# Conflicts:
#	app/code/Magento/CatalogImportExport/Model/Import/Product.php

amenk committed:
#23191 Port delete processing from commit 769c48d239648fd700f5b18157f2a5ef8d726b77 after merge

amenk commented:
Ported changes, next I will add tests

amenk committed:
#23191 Add unit test to check new ability to inject more custom link types + CSFix

amenk commented:
@sidolov I added unit tests to cover the newly exposed functionality (add custom linkToName id).

In the end of the day it's meant to modules like https://github.com/magepal/magento2-link-product/issues/2#issuecomment-463673170 to be able to create custom link types. This cannot be covered in a integration test here (without adding the module to the core), so a unit test suits best here.

The exiting functionality is already covered with integration tests in the import module.

Please review and hopefully merge :-)

amenk committed:
#23191 Fix class name of test

amenk commented:
I have a problem with CyclomaticComplexity of 11 of 

    \Magento\CatalogImportExport\Model\Import\Product\LinkProcessor::processLinkBunches

It's since we added the if() to check for empty links.

I do not find a easy way to reduce it, because the if contains a continue statement.

So I head to add a SuppressWarnings

amenk committed:
#23191 Fix static tests

We have to ignore the CyclomaticComplexity of 11 of

\Magento\CatalogImportExport\Model\Import\Product\LinkProcessor::processLinkBunches

It's since we added the if() to check for empty links.
I do not find a easy way to reduce it, because the if contains a continue statement.

amenk commented:
> I have a problem with CyclomaticComplexity of 11 of

I found a way by externalizing filtering ...

amenk committed:
#23191 Reduce CyclomaticComplexity of processLinkBunches

amenk commented:
Static Tests fail for B2B, not for CE .. how can I fix this locally for B2B ???

https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/23191/2b6abf34e3ee6c73096af27d83615ac2/Statics/allure-report-b2b/index.html

amenk committed:
#23191 Fix static tests for B2B edition, which checks tests

amenk committed:
#23191 Fix static tests: trailing comma in function call

amenk commented:
@sidolov @sivaschenko All tests are green. can this be merged now?

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7035](https://jira.corp.magento.com/browse/ENGCOM-7035) has been created to process this Pull Request
:eight_spoked_asterisk: @ihor-sviziev, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

amenk commented:
@ihor-sviziev Thanks a lot for reviewing & accepting. What are the next steps towards a merge?

ihor-sviziev commented:
@amenk it will go through testing and merging flow.
You can track the status on the following flow https://github.com/orgs/magento/projects/5#card-22569262

amenk commented:
@ihor-sviziev Okay, I think changes-requested by sivaschenko should be cleared, because I added the tests? All changes seem to be resolved. Shall I click "Dismiss review" ?

ihor-sviziev commented:
@amenk no need, your PR already waiting to be tested

engcom-Alfa commented:
:heavy_check_mark: QA Passed
